{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e3d7ec2",
   "metadata": {},
   "source": [
    "# Outline of the metabolomics data analysis\n",
    "\n",
    "In this notebook we briefly go through the steps taken to perform the analysis on the metabolomics dataset in the paper.\n",
    "\n",
    "#### Requirements\n",
    "\n",
    "We recommend creating a fresh python environment and installing the required libraries via\n",
    "```\n",
    "pip install -r requirements.txt\n",
    "```\n",
    "\n",
    "### 0. Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81b42c91",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.io as sio\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from tensorly.decomposition import non_negative_parafac\n",
    "\n",
    "import matcouply\n",
    "from matcouply.decomposition import cmf_aoadmm, initialize_cmf\n",
    "from matcouply.penalties import Parafac2,NonNegativity\n",
    "\n",
    "from copy import deepcopy\n",
    "import plot_funcs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00c3c587",
   "metadata": {},
   "source": [
    "### 1. Load the data and preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77297529",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data\n",
    "# data = sio.loadmat('NMR+Gene+SNP+Metav3_Feb14_2023.mat')[\"Z\"]\n",
    "\n",
    "# data = np.empty(\n",
    "#     (data_object[\"object\"][0].shape[0],\n",
    "#      data_object[\"object\"][0].shape[1],\n",
    "#      K)\n",
    "# )\n",
    "\n",
    "# mask = np.empty_like(data)\n",
    "\n",
    "# for k in range(K):\n",
    "#     data[:,:,k] = data_object[\"object\"][k]\n",
    "#     mask[:,:,k] = data_object[\"miss\"][k]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84966ce2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preprocessing\n",
    "\n",
    "# Remove outliers (if applicable)\n",
    "\n",
    "# Scale data (if applicable)\n",
    "\n",
    "def preprocess_centerscale(X, flag_scale=True, flag_center=True):\n",
    "    \"\"\"\n",
    "    Preprocess a 3D array (subjects × time × metabolites):\n",
    "    - Center across subjects (mode-1)\n",
    "    - Scale within each metabolite (mode-3) using RMS\n",
    "    - Handles NaNs\n",
    "\n",
    "    Parameters:\n",
    "    - X: np.ndarray, shape (subjects, time, metabolites)\n",
    "    - flag_scale: bool, whether to apply RMS scaling\n",
    "    - flag_center: bool, whether to apply centering\n",
    "\n",
    "    Returns:\n",
    "    - Xpre: np.ndarray, preprocessed version of X\n",
    "    \"\"\"\n",
    "    Xpre = np.copy(X)\n",
    "\n",
    "    if flag_center:\n",
    "        mean_subjects = np.nanmean(Xpre, axis=0)\n",
    "        for i in range(Xpre.shape[0]):\n",
    "            Xpre[i, :, :] = Xpre[i, :, :] - mean_subjects\n",
    "\n",
    "    if flag_scale:\n",
    "        for k in range(Xpre.shape[2]):\n",
    "            rms = np.sqrt(np.nanmean(Xpre[:, :, k] ** 2))\n",
    "            if rms != 0:\n",
    "                Xpre[:, :, k] /= rms\n",
    "            else:\n",
    "                Xpre[:, :, k] = 0\n",
    "\n",
    "    return Xpre\n",
    "\n",
    "# data = preprocess_centerscale(train, flag_scale=True, flag_center=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97c4ce39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Missing data\n",
    "\n",
    "# Here imputing with the mean across subjects are a spcific time-point.\n",
    "# This will be later updated while fitting with EM.\n",
    "\n",
    "# for i in range(data.shape[0]):\n",
    "#     for j in range(data.shape[1]):\n",
    "#         for k in range(K):\n",
    "#             if mask[i,j,k] == 0:\n",
    "#                 data[i,j,k] = np.nan\n",
    "\n",
    "# for i in range(data.shape[0]):\n",
    "#     for j in range(data.shape[1]):\n",
    "#         for k in range(K):\n",
    "#             if mask[i,j,k] == 0:\n",
    "#                 data[i,j,k] = np.nanmean(data[i,j,:])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a401cfc",
   "metadata": {},
   "source": [
    "### 2. Reproducibility and replicability analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5bafaaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check uniqueness\n",
    "\n",
    "# CMF\n",
    "\n",
    "# %run uniqueness_check.py cmf 2 0 -m males\n",
    "# %run uniqueness_check.py cmf 3 0 -m males\n",
    "# %run uniqueness_check.py cmf 4 0 -m males\n",
    "# %run uniqueness_check.py cmf 5 0 -m males\n",
    "# %run uniqueness_check.py cmf 6 0 -m males\n",
    "# %run uniqueness_check.py cmf 7 0 -m males\n",
    "# %run uniqueness_check.py cmf 8 0 -m males\n",
    "\n",
    "# %run uniqueness_check.py cmf 2 0 -m females\n",
    "# %run uniqueness_check.py cmf 3 0 -m females\n",
    "# %run uniqueness_check.py cmf 4 0 -m females\n",
    "# %run uniqueness_check.py cmf 5 0 -m females\n",
    "# %run uniqueness_check.py cmf 6 0 -m females\n",
    "# %run uniqueness_check.py cmf 7 0 -m females\n",
    "# %run uniqueness_check.py cmf 8 0 -m females\n",
    "\n",
    "# PARAFAC2\n",
    "\n",
    "# %run uniqueness_check.py parafac2 2 0 -m males\n",
    "# %run uniqueness_check.py parafac2 3 0 -m males\n",
    "# %run uniqueness_check.py parafac2 4 0 -m males\n",
    "# %run uniqueness_check.py parafac2 5 0 -m males\n",
    "# %run uniqueness_check.py parafac2 6 0 -m males\n",
    "# %run uniqueness_check.py parafac2 7 0 -m males\n",
    "# %run uniqueness_check.py parafac2 8 0 -m males\n",
    "\n",
    "# %run uniqueness_check.py parafac2 2 0 -m females\n",
    "# %run uniqueness_check.py parafac2 3 0 -m females\n",
    "# %run uniqueness_check.py parafac2 4 0 -m females\n",
    "# %run uniqueness_check.py parafac2 5 0 -m females\n",
    "# %run uniqueness_check.py parafac2 6 0 -m females\n",
    "# %run uniqueness_check.py parafac2 7 0 -m females\n",
    "# %run uniqueness_check.py parafac2 8 0 -m females\n",
    "\n",
    "# CP\n",
    "\n",
    "# %run uniqueness_check.py cp 2 0 -m males\n",
    "# %run uniqueness_check.py cp 3 0 -m males\n",
    "# %run uniqueness_check.py cp 4 0 -m males\n",
    "# %run uniqueness_check.py cp 5 0 -m males\n",
    "# %run uniqueness_check.py cp 6 0 -m males\n",
    "# %run uniqueness_check.py cp 7 0 -m males\n",
    "# %run uniqueness_check.py cp 8 0 -m males\n",
    "\n",
    "# %run uniqueness_check.py cp 2 0 -m females\n",
    "# %run uniqueness_check.py cp 3 0 -m females\n",
    "# %run uniqueness_check.py cp 4 0 -m females\n",
    "# %run uniqueness_check.py cp 5 0 -m females\n",
    "# %run uniqueness_check.py cp 6 0 -m females\n",
    "# %run uniqueness_check.py cp 7 0 -m females\n",
    "# %run uniqueness_check.py cp 8 0 -m females"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf433e2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check replicability\n",
    "\n",
    "# CMF\n",
    "\n",
    "# %run replicability_check.py cmf 2 0 -m males\n",
    "# %run replicability_check.py cmf 3 0 -m males\n",
    "# %run replicability_check.py cmf 4 0 -m males\n",
    "# %run replicability_check.py cmf 5 0 -m males\n",
    "# %run replicability_check.py cmf 6 0 -m males\n",
    "# %run replicability_check.py cmf 7 0 -m males\n",
    "# %run replicability_check.py cmf 8 0 -m males\n",
    "\n",
    "# %run replicability_check.py cmf 2 0 -m females\n",
    "# %run replicability_check.py cmf 3 0 -m females\n",
    "# %run replicability_check.py cmf 4 0 -m females\n",
    "# %run replicability_check.py cmf 5 0 -m females\n",
    "# %run replicability_check.py cmf 6 0 -m females\n",
    "# %run replicability_check.py cmf 7 0 -m females\n",
    "# %run replicability_check.py cmf 8 0 -m females\n",
    "\n",
    "# PARAFAC2\n",
    "\n",
    "# %run replicability_check.py parafac2 2 0 -m males\n",
    "# %run replicability_check.py parafac2 3 0 -m males\n",
    "# %run replicability_check.py parafac2 4 0 -m males\n",
    "# %run replicability_check.py parafac2 5 0 -m males\n",
    "# %run replicability_check.py parafac2 6 0 -m males\n",
    "# %run replicability_check.py parafac2 7 0 -m males\n",
    "# %run replicability_check.py parafac2 8 0 -m males\n",
    "\n",
    "# %run replicability_check.py parafac2 2 0 -m females\n",
    "# %run replicability_check.py parafac2 3 0 -m females\n",
    "# %run replicability_check.py parafac2 4 0 -m females\n",
    "# %run replicability_check.py parafac2 5 0 -m females\n",
    "# %run replicability_check.py parafac2 6 0 -m females\n",
    "# %run replicability_check.py parafac2 7 0 -m females\n",
    "# %run replicability_check.py parafac2 8 0 -m females\n",
    "\n",
    "# CP\n",
    "\n",
    "# %run replicability_check.py cp 2 0 -m males\n",
    "# %run replicability_check.py cp 3 0 -m males\n",
    "# %run replicability_check.py cp 4 0 -m males\n",
    "# %run replicability_check.py cp 5 0 -m males\n",
    "# %run replicability_check.py cp 6 0 -m males\n",
    "# %run replicability_check.py cp 7 0 -m males\n",
    "# %run replicability_check.py cp 8 0 -m males\n",
    "\n",
    "# %run replicability_check.py cp 2 0 -m females\n",
    "# %run replicability_check.py cp 3 0 -m females\n",
    "# %run replicability_check.py cp 4 0 -m females\n",
    "# %run replicability_check.py cp 5 0 -m females\n",
    "# %run replicability_check.py cp 6 0 -m females\n",
    "# %run replicability_check.py cp 7 0 -m females\n",
    "# %run replicability_check.py cp 8 0 -m females"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60b1da03",
   "metadata": {},
   "source": [
    "The output of these scripts is located at `Metabolomics/results/uniqueness/` and `Metabolomics/results/replicability/`, respectively. The results can be nicely summarized in a plot by `uniqueness_analysis_metabolomics.ipynb` and `replicability_analysis_metabolomics.ipynb`. To reproduce the exact plot of the paper `paper_plots.ipynb` can be used."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab2469e3",
   "metadata": {},
   "source": [
    "### 3. Model fitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e8d1329",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fitting CP models\n",
    "\n",
    "best_error = np.inf\n",
    "best_factors = None\n",
    "\n",
    "for init in range(40):\n",
    "\n",
    "    factors, errors = non_negative_parafac(\n",
    "        tensor=data,\n",
    "        tol=1e-8,\n",
    "        rank=6,\n",
    "        return_errors=True,\n",
    "        n_iter_max=15000,\n",
    "        mask=mask,\n",
    "    )\n",
    "\n",
    "    if len(errors) < 14990: # If we have not reached max number of iters\n",
    "        if errors[-1] < best_error: # did we get a better loss?\n",
    "            best_error = errors[-1]\n",
    "            best_factors = factors[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b24e708",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fitting PARAFAC2\n",
    "\n",
    "best_error = np.inf\n",
    "best_factors = None\n",
    "\n",
    "for init in range(40):\n",
    "\n",
    "    (weights, (D, B, A)), run_diagnostics = cmf_aoadmm(\n",
    "        matrices=data.T, # This needs to be transposed due to ordering of factors in matcouply\n",
    "        rank=6,\n",
    "        return_errors=True,\n",
    "        n_iter_max=15000,\n",
    "        regs=[\n",
    "            [NonNegativity()],\n",
    "            [NonNegativity(), Parafac2()],\n",
    "            [NonNegativity()],\n",
    "        ],\n",
    "        tol=1e-8,\n",
    "        inner_n_iter_max=20,\n",
    "        absolute_tol=1e-6,\n",
    "        feasibility_tol=1e-5,\n",
    "        inner_tol=1e-5,\n",
    "        mask=mask,\n",
    "    )\n",
    "\n",
    "    if (\n",
    "        len(run_diagnostics.regularized_loss) > 14990\n",
    "    ):\n",
    "        continue\n",
    "    else:\n",
    "        if run_diagnostics.regularized_loss[-1] < best_error:\n",
    "            best_error = run_diagnostics.regularized_loss[-1]\n",
    "            best_factors = (D, B, A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d959c80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fitting CMF models\n",
    "\n",
    "best_error = np.inf\n",
    "best_factors = None\n",
    "\n",
    "for init in range(40):\n",
    "\n",
    "    weights, (A_init, B_init, D_init) = initialize_cmf(\n",
    "        matrices=data.T,\n",
    "        rank=6,\n",
    "        init=\"random\",\n",
    "        svd_fun=\"truncated_svd\",\n",
    "    )\n",
    "\n",
    "    A_init = np.ones_like(A_init)\n",
    "    cmf_init = weights, (A_init, B_init, D_init)\n",
    "\n",
    "    (weights, (D, B, A)), run_diagnostics = cmf_aoadmm(\n",
    "        matrices=data.T, # This needs to be transposed due to ordering of factors in matcouply\n",
    "        rank=6,\n",
    "        return_errors=True,\n",
    "        n_iter_max=8000,\n",
    "        regs=[[NonNegativity()], [NonNegativity()], [NonNegativity()]],\n",
    "        tol=1e-8,\n",
    "        init=cmf_init,\n",
    "        update_A=False,\n",
    "        inner_n_iter_max=20,\n",
    "        absolute_tol=1e-6,\n",
    "        feasibility_tol=1e-5,\n",
    "        inner_tol=1e-5,\n",
    "        mask=mask,\n",
    "    )\n",
    "\n",
    "    if (\n",
    "        len(run_diagnostics.regularized_loss) > 7990\n",
    "    ):\n",
    "        continue\n",
    "    else:\n",
    "        if run_diagnostics.regularized_loss[-1] < best_error:\n",
    "            best_error = run_diagnostics.regularized_loss[-1]\n",
    "            best_factors = (D, B, A)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c7f5d83",
   "metadata": {},
   "source": [
    "### 4. Plot results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "737fe273",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata\n",
    "\n",
    "# ⚠️ Load these for males\n",
    "# df = pd.read_csv(\"Metabolomics/meta_males_safe.csv\")\n",
    "# group = df['group'].astype(\"string\").to_numpy()  \n",
    "\n",
    "# metabolites = ['Total-C', 'Remnant-C', 'VLDL-C', 'LDL-C', 'HDL-C', 'Total-TG', 'VLDL-TG', 'LDL-TG', 'HDL-TG', 'Total-PL', 'VLDL-PL', 'LDL-PL', 'HDL-PL', 'Total-CE', 'VLDL-CE', 'LDL-CE', 'HDL-CE', 'Total-FC', 'VLDL-FC', 'LDL-FC', 'HDL-FC', 'Total-L', 'VLDL-L', 'LDL-L', 'HDL-L', 'Total-P', 'VLDL-P', 'LDL-P', 'HDL-P', 'VLDL size', 'LDL size', 'HDL size', 'ApoB', 'ApoA1', 'Total-FA', 'Unsaturation', 'Omega-3', 'Omega-6', 'PUFA', 'MUFA', 'SFA', 'LA', 'DHA', 'Ala', 'Gln', 'Gly', 'His', 'Total BCAA', 'Ile', 'Leu', 'Val', 'Tyr', 'Glucose', 'Lactate', 'Pyruvate', 'Citrate', 'bOHbutyrate', 'Acetate', 'Acetoacetate', 'Acetone', 'GlycA', 'XXL-VLDL-P', 'XXL-VLDL-L', 'XXL-VLDL-PL', 'XXL-VLDL-C', 'XXL-VLDL-CE', 'XXL-VLDL-FC', 'XXL-VLDL-TG', 'XL-VLDL-P', 'XL-VLDL-L', 'XL-VLDL-PL', 'XL-VLDL-C', 'XL-VLDL-CE', 'XL-VLDL-FC', 'XL-VLDL-TG', 'L-VLDL-P', 'L-VLDL-L', 'L-VLDL-PL', 'L-VLDL-C', 'L-VLDL-CE', 'L-VLDL-FC', 'L-VLDL-TG', 'M-VLDL-P', 'M-VLDL-L', 'M-VLDL-PL', 'M-VLDL-C', 'M-VLDL-CE', 'M-VLDL-FC', 'M-VLDL-TG', 'S-VLDL-P', 'S-VLDL-L', 'S-VLDL-PL', 'S-VLDL-C', 'S-VLDL-CE', 'S-VLDL-FC', 'S-VLDL-TG', 'XS-VLDL-P', 'XS-VLDL-L', 'XS-VLDL-PL', 'XS-VLDL-C', 'XS-VLDL-CE', 'XS-VLDL-FC', 'XS-VLDL-TG', 'IDL-P', 'IDL-L', 'IDL-PL', 'IDL-C', 'IDL-CE', 'IDL-FC', 'IDL-TG', 'L-LDL-P', 'L-LDL-L', 'L-LDL-PL', 'L-LDL-C', 'L-LDL-CE', 'L-LDL-FC', 'L-LDL-TG', 'M-LDL-P', 'M-LDL-L', 'M-LDL-PL', 'M-LDL-C', 'M-LDL-CE', 'M-LDL-FC', 'M-LDL-TG', 'S-LDL-P', 'S-LDL-L', 'S-LDL-PL', 'S-LDL-C', 'S-LDL-CE', 'S-LDL-FC', 'S-LDL-TG', 'XL-HDL-P', 'XL-HDL-L', 'XL-HDL-PL', 'XL-HDL-C', 'XL-HDL-CE', 'XL-HDL-FC', 'XL-HDL-TG', 'L-HDL-P', 'L-HDL-L', 'L-HDL-PL', 'L-HDL-C', 'L-HDL-CE', 'L-HDL-FC', 'L-HDL-TG', 'M-HDL-P', 'M-HDL-L', 'M-HDL-PL', 'M-HDL-C', 'M-HDL-CE', 'M-HDL-FC', 'M-HDL-TG', 'S-HDL-P', 'S-HDL-L', 'S-HDL-PL', 'S-HDL-C', 'S-HDL-CE', 'S-HDL-FC', 'S-HDL-TG', 'Insulin', 'Cpeptid']\n",
    "# metabolite_class = ['Cholesterol', 'Cholesterol', 'Cholesterol', 'Cholesterol', 'Cholesterol', 'Triglycerides', 'Triglycerides', 'Triglycerides', 'Triglycerides', 'Phospholipids', 'Phospholipids', 'Phospholipids', 'Phospholipids', 'Cholesteryl esters', 'Cholesteryl esters', 'Cholesteryl esters', 'Cholesteryl esters', 'Free cholesterol', 'Free cholesterol', 'Free cholesterol', 'Free cholesterol', 'Total lipids', 'Total lipids', 'Total lipids', 'Total lipids', 'Lipoprotein particle concentrations', 'Lipoprotein particle concentrations', 'Lipoprotein particle concentrations', 'Lipoprotein particle concentrations', 'Lipoprotein particle sizes', 'Lipoprotein particle sizes', 'Lipoprotein particle sizes', 'Apolipoproteins', 'Apolipoproteins', 'Fatty acids', 'Fatty acids', 'Fatty acids', 'Fatty acids', 'Fatty acids', 'Fatty acids', 'Fatty acids', 'Fatty acids', 'Fatty acids', 'Amino acids', 'Amino acids', 'Amino acids', 'Amino acids', 'Amino acids', 'Amino acids', 'Amino acids', 'Amino acids', 'Amino acids', 'Glycolysis related metabolites', 'Glycolysis related metabolites', 'Glycolysis related metabolites', 'Glycolysis related metabolites', 'Ketone bodies', 'Ketone bodies', 'Ketone bodies', 'Ketone bodies', 'Inflammation', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', 'Lipoprotein subclasses', '', '']\n",
    "\n",
    "# # find indices where metabolite_class is 'Ketone bodies'\n",
    "# ketone_indices = [i for i, x in enumerate(metabolite_class) if x == 'Ketone bodies']\n",
    "# ketone_metabolites = [metabolites[i] for i in ketone_indices]\n",
    "\n",
    "# glycolisis = [i for i, x in enumerate(metabolite_class) if x == 'Glycolysis related metabolites']\n",
    "# glycolisis_metabolites = [metabolites[i] for i in glycolisis]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "912d46a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = 'cmf' # can be 'cmf', 'cp' or 'parafac2'\n",
    "n_components = 6\n",
    "\n",
    "A = best_factors[2]\n",
    "C = best_factors[0]\n",
    "if model != 'cp':\n",
    "    B = best_factors[1]\n",
    "else:\n",
    "    B = [deepcopy(best_factors[1]) for _ in range(A.shape[0])]\n",
    "\n",
    "if model == 'cp': # if model is 'cp' the order of the factors needs to be changed\n",
    "    (A,B,C) = (C,B,A)\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    n_components, 3,\n",
    "    figsize=(8,0.65 * 8),\n",
    "    constrained_layout=False\n",
    ")\n",
    "\n",
    "fig.subplots_adjust(\n",
    "    left=0.06, right=0.935, bottom=0.08, top=0.97,\n",
    "    wspace=0.24, hspace=0.12\n",
    ")\n",
    "\n",
    "for i, comp in enumerate(range(6)): # keep the same ordering as in the paper\n",
    "    if i == 0: \n",
    "        plot_funcs.plot_metabolite_component(factors=(A,B,C),comp=comp, metabolites=metabolites,axes=axs[i,0],plot_title=True,comp_i=i)\n",
    "        plot_funcs.plot_metabolomics_profile_component(factors=(A,B,C),comp=comp,axes=axs[i,1],comp_i=i,plot_title=True,time_points=[0,0.25,0.5,1,1.5,2,2.5,4])\n",
    "        plot_funcs.plot_metabolomics_stratified_profiles_component(factors=(A,B,C),comp=comp,axes=axs[i,2],plot_title=True,group=group,time_points=[0,0.25,0.5,1,1.5,2,2.5,4])\n",
    "    elif i == 5: \n",
    "        plot_funcs.plot_metabolite_component(factors=(A,B,C),comp=comp,metabolites=metabolites,axes=axs[i,0],plot_xticks=True,plot_legend=True,comp_i=i)\n",
    "        plot_funcs.plot_metabolomics_profile_component(factors=(A,B,C),comp=comp,axes=axs[i,1],comp_i=i,plot_legend=True,plot_xticks=True,time_points=[0,0.25,0.5,1,1.5,2,2.5,4])\n",
    "        plot_funcs.plot_metabolomics_stratified_profiles_component(factors=(A,B,C),comp=comp,axes=axs[i,2],plot_legend=True,plot_xticks=True,group=group,time_points=[0,0.25,0.5,1,1.5,2,2.5,4])\n",
    "    else: \n",
    "        plot_funcs.plot_metabolite_component(factors=(A,B,C),comp=comp,metabolites=metabolites,axes=axs[i,0],comp_i=i)\n",
    "        plot_funcs.plot_metabolomics_profile_component(factors=(A,B,C),comp=comp,axes=axs[i,1],comp_i=i,time_points=[0,0.25,0.5,1,1.5,2,2.5,4])\n",
    "        plot_funcs.plot_metabolomics_stratified_profiles_component(factors=(A,B,C),comp=comp,axes=axs[i,2],group=group,time_points=[0,0.25,0.5,1,1.5,2,2.5,4])\n",
    "\n",
    "\n",
    "axs[0,0].set_xticks([0,49,99,149])\n",
    "axs[1,0].set_xticks([0,49,99,149])\n",
    "axs[2,0].set_xticks([0,49,99,149])\n",
    "axs[3,0].set_xticks([0,49,99,149])\n",
    "axs[4,0].set_xticks([0,49,99,149])\n",
    "axs[5,0].set_xticks([0,49,99,149])\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "indiv-time-profiles-env (3.12.3)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
